scylla/src/transport/session.rs

```diff
@@ -1200,13 +1200,15 @@ impl Session {
             let first_serialized_value =
                 values.batch_values_iter().next_serialized().transpose()?;
 
-            match (first_serialized_value, batch.statements.first()) {
+            // The temporary "p" is necessary because lifetimes
+            let p = match (first_serialized_value, batch.statements.first()) {
                 (Some(first_serialized_value), Some(BatchStatement::PreparedStatement(ps))) => {
                     let token = ps.calculate_token(&first_serialized_value)?;
                     (Some(first_serialized_value), token, ps.get_keyspace_name())
                 }
                 _ => (None, None, None),
-            }
+            };
+            p
```

Comment:
It's interesting that this would cause issues, I thought both versions should have identical semantics.

Reply:
Also, if the problem happens with the new API but not with old, doesn't it mean there is possible some issue within new API definitions?

Reply:
> It's interesting that this would cause issues, I thought both versions should have identical semantics.

In the version of the PR that you reviewed, `BatchValueIterator` would lend its value. I suspect that could have an effect here. I'm not sure it is an issue anymore.

> Also, if the problem happens with the new API but not with old, doesn't it mean there is possible some issue within new API definitions?

Well, I saw that there is an existing test for batches:

```rust
#[test]
fn check_batch_values_iterator_is_not_lending() {
    // This is an interesting property if we want to improve the batch shard selection heuristic
    fn f(bv: impl LegacyBatchValues) {
        let mut it = bv.batch_values_iter();
        let mut it2 = bv.batch_values_iter();
        // Make sure we can hold all these at the same time
        let v = vec![
            it.next_serialized().unwrap().unwrap(),
            it2.next_serialized().unwrap().unwrap(),
            it.next_serialized().unwrap().unwrap(),
            it2.next_serialized().unwrap().unwrap(),
        ];
        let _ = v;
    }
    f(((10,), (11,)))
}
```

On the other hand, I don't see _why_ it was important for the iterator not to lend (I couldn't infer it from the comments, GitHub discussions, commit descriptions etc.). Nevertheless, In the latest version of the PR the `BatchValuesIterator` does not lend (it's not really a true iterator anymore), so the issue is not there and I adapter the test successfully.